envoy/http/header_evaluator.h

```diff
@@ -13,12 +13,16 @@ class HeaderEvaluator {
   /**
    * Apply the header operations that are saved in the HeaderEvaluator. An example of the operation
    * is to add a new header name `foo` to the target header map and the header value is extracted
-   * from the `bar` field in the stream_info.
+   * from the `bar` field in the stream_info, request headers or response headers.
    *
    * @param headers the target header map to be mutated.
+   * @param request_headers request headers to be used in the header manipulation.
+   * @param response_headers response headers to be used in the header manipulation.
    * @param stream_info the source of values that can be used in the evaluation.
    */
   virtual void evaluateHeaders(Http::HeaderMap& headers,
+                               const Http::RequestHeaderMap& request_headers,
+                               const Http::ResponseHeaderMap& response_headers,
                                const StreamInfo::StreamInfo& stream_info) const PURE;
```

Comment:
Could we keep both the original version and new version methods in this PR to minimize the impact on existing code.

Reply:
Unfortunately this is not possible without further complicating the code. `evaluateHeaders` method simply iterates over headers and invokes formatters. The problem is that `evaluateHeaders` does not know if the formatters are "old" style or the "new" style. The new (access log compatible) formatters need response headers to evaluate RESP(X) formatter, so they must be passed as params. If it happens that underlying formatters are "old" style, the call to them goes through [HttpHeaderFormatterBridge](https://github.com/envoyproxy/envoy/blob/17a81b5cddff3f280624c4d42abf8e50787ea77d/source/common/router/header_formatter.h#L128-L147), which simply ignores `request_headers` and `response_headers` and passes only stream_info to the formatter. That in my opinion was the least complicated way to have "old" and "new" formatters and fork between them based on runtime guard. The runtime guard controls a path when config is received. It either forwards config to "old" parser which creates "old" formatters or forwards the config to a "new" (access log's) parser  which creates "new" (actually access log's formatters). After that, during actual formatting, the code is the same. If I kept original version of `evaluateHeaders` I would have to check runtime guard each time `evaluateHeaders` is called, which would complicate code and affect performance.

I understand that this PR is large and code reviewing it is not trivial and I truly appreciate your effort to understand the changes.

Reply:
What I mean by `both the original version and new version methods` is to keep the interface rather than the implementation.

We can keep an old version interface to wrapper the new implementation like this:

```
void evaluateHeaders(Http::HeaderMap& headers, const StreamInfo::StreamInfo& stream_info) const override {
       evaluateHeaders(headers, stream_info->getRequestHeaders() != nullptr 
          ? *stream_info->getRequestHeaders() : *Http::StaticEmptyHeaders::get().request_headers, *Http::StaticEmptyHeaders::get().response_headers, stream_info);
}
``` 

By this simple wrapper, we can reduce much of the code change to make this PR more clean.

Reply:
This wrapper indeed maps 2 parameters to 4 parameters version, but I do not see how can we use it. We need an opposite operation: take 4 parameters and reduce it to 2 (when runtime guard created "old" style formatters) and this is exactly what is done in [HttpHeaderFormatterBridge.](https://github.com/envoyproxy/envoy/pull/21932#:~:text=HttpHeaderFormatterBridge)